File: 215. Kth Largest Element in an Array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findKthLargest = function(nums, k) {
    if ( nums === null || nums.length === 0 ) {
        return null;
    }

    quickSort(nums, 0, nums.length - 1);

    return nums[nums.length-k];
};

var quickSort = function(nums, left, right) {

    if ( left >= right ) {
        return;
    }

    var l = left;
    var r = right;
    var tmp
    var pivot = nums[Math.floor((left+right)/2)];

    while ( l <= r ) {
        while ( l <= r && nums[l] < pivot ) {
            l++;
        }
        while ( l <= r && nums[r] > pivot ) {
            r--
        }
        if ( l <= r ) {
            tmp = nums[l];
            nums[l] = nums[r];
            nums[r] = tmp;

            l++;
            r--;
        }
    }

    quickSort(nums, left, r);
    quickSort(nums, l, right);
};

File: 215. Kth Largest Element in an Array.py
class Solution(object):
    def findKthLargest(self, nums, k):
        self.quickSort(nums, 0, len(nums) - 1, k)
        return nums[-k]

    def quickSort(self, A, start, end, k):
        if start >= end:
            return

        left = start
        right = end
        pivot = A[(start + end) / 2]

        while left <= right:
            while left <= right and A[left] < pivot:
                left += 1
            while left <= right and A[right] > pivot:
                right -= 1
            if left <= right:
                A[left], A[right] = A[right], A[left]
                left += 1
                right -= 1
        if left <= len(A) - k:
            self.quickSort(A, left, end, k)
        else:
            self.quickSort(A, start, right, k)
